tests/test_optimizer/test_dist_adafactor.py

```diff
@@ -0,0 +1,527 @@
+import pytest
+import os
+import sys
+import copy
+import torch
+import torch.distributed as dist
+from torch import nn
+from torch.testing import assert_close
+
+import colossalai
+from colossalai.tensor.d_tensor.layout import Layout
+from colossalai.tensor.d_tensor import layout_converter
+from colossalai.tensor.d_tensor.sharding_spec import DimSpec
+from colossalai.testing import parameterize, rerun_if_address_is_in_use, spawn
+from colossalai.utils import set_seed
+from colossalai.device.device_mesh import DeviceMesh
+from colossalai.cluster import DistCoordinator, ProcessGroupMesh
+from colossalai.tensor.d_tensor import (
+    is_distributed_tensor,
+    distribute_tensor,
+    sharded_tensor_to_param,
+    shard_rowwise,
+    shard_colwise,
+    get_layout,
+    get_sharding_spec
+)
+from colossalai.tensor.d_tensor.api import clear_layout_converter
+from colossalai.shardformer.layer._operation import _gather
+from colossalai.shardformer.layer import Linear1D_Col, Linear1D_Row
+from colossalai.zero import LowLevelZeroOptimizer
+from colossalai.nn.optimizer.adafactor import Adafactor
+from colossalai.nn.optimizer.distributed_adafactor import DistributedAdaFactor
+from colossalai.booster import Booster
+from colossalai.booster.plugin import TorchDDPPlugin
+
+HEIGHT = 4096
+WIDTH = 4096
+_TP_SPEC = DimSpec([0])
+
+def init_dist():
+    rank = int(os.environ['RANK']) 
+    local_rank = int(os.environ['LOCAL_RANK'])
+    world_size = int(os.environ['WORLD_SIZE'])
+    dist.init_process_group(world_size=world_size, rank=rank,
+                            init_method="env://", backend="nccl")
+    torch.cuda.set_device(local_rank)
+
+def correctness_verify(tensor1: torch.Tensor, tensor2: torch.Tensor, dtype: torch.dtype = torch.float32):
+    rtol = None
+    atol = None
+    if dtype is torch.float32:
+        rtol = 1e-05
+        atol = 1e-05
+    elif dtype is torch.float16:
+        rtol = 5e-2
+        atol = 5e-4
+    elif dtype is torch.bfloat16:
+        rtol = 4e-3
+        atol = 4e-3
+    
+    return torch.all(tensor1.isclose(tensor2, rtol=rtol, atol=atol, equal_nan=True))
```

Comment:
Nan is not acceptable

Reply:
> Nan is not acceptable


Fixed in https://github.com/hpcaitech/ColossalAI/pull/5484/commits/acb2148512fe31c40dc4e0a018c2d372a05d670a .